### PR TITLE
Jwoo/ios bleed 2

### PR DESCRIPTION
**Pull request by nesalang**
1. ios renderer implementation
2. please note that shared code should not be reviewed as it will be behind.
Simulator Screen Shot - iPhone 6s - 2019-03-15 at 16 03 59
Simulator Screen Shot - iPhone 6s - 2019-03-15 at 16 03 50
Simulator Screen Shot - iPhone 6s - 2019-03-15 at 16 03 45

### CODE REVIEW

**Comment by paulcam206** [comment]
Looks like you need to update the dotnet tests:

```
Failed   TestAllTestCards
Error Message:
 Test method AdaptiveCards.Test.AllPayloadTests.TestAllTestCards threw exception: 
System.AggregateException: One or more errors occurred. (Payload file failed: Nested.ColumnSet.Bleed.json) ---> System.Exception: Payload file failed: Nested.ColumnSet.Bleed.json ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.AreEqual failed. Expected:<0>. Actual:<1>. 
Stack Trace:
   at AdaptiveCards.Test.AllPayloadTests.TestPayloadsInDirectory(String path, String[] excludedCards) in C:\BA\6880\s\source\dotnet\test\AdaptiveCards.Test\AllPayloadTests.cs:line 68
--- End of inner exception stack trace ---
    at AdaptiveCards.Test.AllPayloadTests.TestPayloadsInDirectory(String path, String[] excludedCards) in C:\BA\6880\s\source\dotnet\test\AdaptiveCards.Test\AllPayloadTests.cs:line 90
   at AdaptiveCards.Test.AllPayloadTests.TestAllTestCards() in C:\BA\6880\s\source\dotnet\test\AdaptiveCards.Test\AllPayloadTests.cs:line 110
```

File: samples/Tests/Nested.ColumnSet.Bleed.Complex.json [hunk 1]
```diff
@@ -0,0 +1,154 @@
+{
+    "type": "AdaptiveCard",
+    "body": [
+        {
+            "type": "TextBlock",
+            "text": "Only inner left and right most containers bleed to edges.\nNotice that column with Y has longer width",
+            "wrap": true
```
Comment (paulcam206) [inline samples/Tests/Nested.ColumnSet.Bleed.Complex.json]: json files should be indented with tabs. have you [configured your editor](https://editorconfig.org/#download) to apply our formatting?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm [hunk 2]
```diff
@@ -6,10 +6,104 @@
...
+                    // 2. bv is added to bleed target view (tv), which is also a parent view.
+                    // bv is then pinned to the tv according to the bleed direction
+                    // bv gets current container view's (cv) container style
+                    // and cv's container style is reset to transparent, such that
+                    // bv's container style will be diplayed.
+                    // container view's stack view (csv) holds content views, and bv dislpays container style
+                    // we transpose them, and get the final result
+                    UIView *backgroundView = [[UIView alloc] init];
+                    container.backgroundView = backgroundView;
+                    backgroundView.translatesAutoresizingMaskIntoConstraints = NO;
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm]: >               [](start = 83, length = 14)

nitpick: trailing spaces

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm [hunk 3]
```diff
@@ -6,10 +6,104 @@
...
+                    backgroundView.translatesAutoresizingMaskIntoConstraints = NO;                   
+                    [view addSubview:backgroundView];
+                    [view sendSubviewToBack:backgroundView];
+                    backgroundView.backgroundColor = container.backgroundColor;
+                    container.backgroundColor = UIColor.clearColor;
+                    [backgroundView.topAnchor constraintEqualToAnchor:container.topAnchor].active = YES;
+                    [backgroundView.bottomAnchor constraintEqualToAnchor:container.bottomAnchor].active = YES;
+
+                    // reset current width constraints and update them with new one according to bleed specification
+                    [container bleed:config->GetSpacing().paddingSpacing priority:1000 target:backgroundView
```
Comment (paulcam206) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm]: >priority:1000  [](start = 73, length = 14)

do we ever `bleed` with a priority other than `1000`?

File: source/shared/cpp/ObjectModel/CollectionTypeElement.cpp [hunk 4]
```diff
@@ -70,22 +71,18 @@ void CollectionTypeElement::ConfigBleed(const AdaptiveCards::ParseContext& conte
 {
     // we allows bleed when self has padding and at least one parent has padding
     const AdaptiveSharedNamespace::InternalId id = context.PaddingParentInternalId();
-    const bool canBleed = GetBleed() && (id != AdaptiveSharedNamespace::InternalId::Invalid);
-    if (canBleed)
+    bool valid = (id != AdaptiveSharedNamespace::InternalId::Invalid);
```
Comment (paulcam206) [inline source/shared/cpp/ObjectModel/CollectionTypeElement.cpp]: >bool [](start = 4, length = 4)

could be `const`

File: source/shared/cpp/ObjectModel/CollectionTypeElement.h [hunk 5]
```diff
@@ -57,18 +66,21 @@ namespace AdaptiveSharedNamespace
...
 
     private:
-        void SetCanBleed(const bool value);
+        void SetCanBleed(const bool value) { m_canBleed = value;}
 
         // Applies padding flag When appropriate
         void ConfigPadding(const AdaptiveCards::ParseContext& context);
         // Applies bleed flag when appropriate
         void ConfigBleed(const AdaptiveCards::ParseContext& context);
+        void SetBleedDirection(const ContainerBleedDirection bleedDirection) { m_bleedDirection= bleedDirection; }
```
Comment (paulcam206) [inline source/shared/cpp/ObjectModel/CollectionTypeElement.h]: >n= [](start = 94, length = 2)

nitpick: should be a space before `=`

File: source/shared/cpp/ObjectModel/ColumnSet.cpp [hunk 6]
```diff
@@ -47,11 +47,84 @@ std::shared_ptr<BaseCardElement> ColumnSetParser::Deserialize(ParseContext& cont
...
+
+    const size_t elemSize = elementArray.size();
+    const size_t endElemIndex = elemSize - 1;
+    elements.reserve(elemSize);
+
+    size_t currentIndex = 0;
+
+    // Child's bleed get more restricted as we walk down the tree
+    // until padding is reset
+    // padding get rested when padding is added to a collection type element
```
Comment (paulcam206) [inline source/shared/cpp/ObjectModel/ColumnSet.cpp]: >get rested [](start = 15, length = 10)

typo: `gets reset`?

File: source/shared/cpp/ObjectModel/ColumnSet.cpp [hunk 7]
```diff
@@ -47,11 +47,84 @@ std::shared_ptr<BaseCardElement> ColumnSetParser::Deserialize(ParseContext& cont
...
+                }
+            } else
+            {
+                context.SetBleedDirection(ContainerBleedDirection::BleedRestricted);
+            }
+        }
+
+        InternalId::Next();        
+        // Parse the element
+        auto el = Column::Deserialize(context, curJsonValue);
```
Comment (paulcam206) [inline source/shared/cpp/ObjectModel/ColumnSet.cpp]: I think part of the problem here is that we shouldn't hardcode the assumption that everything under the `columns` key is, in fact, a column. there could be a custom element in the json. we should be going through the element parser registration to parse. if we do that, we shouldn't have to manually move the internal id.

File: samples/Tests/Nested.ColumnSet.Bleed.Complex.json (hunk 1 above)
Comment (nesalang) [reply samples/Tests/Nested.ColumnSet.Bleed.Complex.json]: it's configured on wsl. i might have to update it on mac

---
In reply to: [268269359](https://github.com/Microsoft/AdaptiveCards/pull/2551#discussion_r268269359) [](ancestors = 268269359)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm (hunk 3 above)
Comment (nesalang) [reply source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/Util.mm]: 1000 is the max value.

---
In reply to: [268272608](https://github.com/Microsoft/AdaptiveCards/pull/2551#discussion_r268272608) [](ancestors = 268272608)

File: source/shared/cpp/ObjectModel/ColumnSet.cpp (hunk 7 above)
Comment (nesalang) [reply source/shared/cpp/ObjectModel/ColumnSet.cpp]: that sounds good, but we don't have column parser registered. should we register it?

---
In reply to: [268278405](https://github.com/Microsoft/AdaptiveCards/pull/2551#discussion_r268278405) [](ancestors = 268278405)

File: source/shared/cpp/ObjectModel/ColumnSet.cpp (hunk 6 above)
Comment (nesalang) [reply source/shared/cpp/ObjectModel/ColumnSet.cpp]: :-) thanks

---
In reply to: [268274205](https://github.com/Microsoft/AdaptiveCards/pull/2551#discussion_r268274205) [](ancestors = 268274205)

**Comment by paulcam206** [comment]
Hey, looks like the new Android CI build caught a real bug! :)

```
/home/vsts/work/1/s/source/android/adaptivecards/src/main/cpp/objectmodel_wrap.cpp:13763:12: error: no matching function for call to 'Deserialize'
    result = AdaptiveCards::Column::Deserialize(*arg1,(Json::Value const &)*arg2);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../../../../../../shared/cpp/ObjectModel/CollectionTypeElement.h:92:47: note: candidate template ignored: couldn't infer template argument 'T'
      std::shared_ptr<T> CollectionTypeElement::Deserialize(ParseContext& context, const Json::Value& value)
                                                ^
  /home/vsts/work/1/s/source/android/adaptivecards/src/main/cpp/objectmodel_wrap.cpp:13793:35: error: no member named 'DeserializeFromString' in 'AdaptiveCards::Column'
    result = AdaptiveCards::Column::DeserializeFromString(*arg1,(std::string const &)*arg2);
             ~~~~~~~~~~~~~~~~~~~~~~~^
```

I'm guessing that you need to run swig :)

**Pull request merged**